packages/client/lib/miner/pendingBlock.ts

```diff
@@ -0,0 +1,171 @@
+  config: Config
+  txPool: TxPool
+  pendingPayloads: [payloadId: Buffer, builder: BlockBuilder][] = []
+
+  constructor(opts: PendingBlockOpts) {
+    this.config = opts.config
+    this.txPool = opts.txPool
+  }
+
+  /**
+   * Starts building a pending block with the given payload
+   * @returns an 8-byte payload identifier to call {@link BlockBuilder.build} with
+   */
+  async start(vm: VM, parentBlock: Block, headerData: Partial<HeaderData> = {}) {
+    const number = parentBlock.header.number.addn(1)
+    const { gasLimit } = parentBlock.header
+    const baseFeePerGas = parentBlock.header.calcNextBaseFee()
+
+    // Set the state root to ensure the resulting state
+    // is based on the parent block's state
+    await vm.stateManager.setStateRoot(parentBlock.header.stateRoot)
+
+    const td = await vm.blockchain.getTotalDifficulty(parentBlock.hash())
+    vm._common.setHardforkByBlockNumber(parentBlock.header.number, td)
+
+    const builder = await vm.buildBlock({
+      parentBlock,
+      headerData: {
+        ...headerData,
+        number,
+        gasLimit,
+        baseFeePerGas,
+      },
+      blockOpts: {
+        putBlockIntoBlockchain: false,
+      },
+    })
+
+    const payloadId = randomBytes(8)
+    this.pendingPayloads.push([payloadId, builder])
+
+    // Add current txs in pool
+    const txs = await this.txPool.txsByPriceAndNonce(vm.stateManager, baseFeePerGas)
+    this.config.logger.info(
+      `Pending: Assembling block from ${txs.length} eligible txs (baseFee: ${baseFeePerGas})`
+    )
+    let index = 0
+    let blockFull = false
+    while (index < txs.length && !blockFull) {
+      try {
+        await builder.addTransaction(txs[index])
+      } catch (error: any) {
+        if (error.message === 'tx has a higher gas limit than the remaining gas in the block') {
+          if (builder.gasUsed.gt(gasLimit.subn(21000))) {
+            // If block has less than 21000 gas remaining, consider it full
+            blockFull = true
+            this.config.logger.info(
+              `Pending: Assembled block full (gasLeft: ${gasLimit.sub(builder.gasUsed)})`
+            )
+          }
+        } else {
+          // If there is an error adding a tx, it will be skipped
+          this.config.logger.debug(
+            `Pending: Skipping tx 0x${txs[index]
+              .hash()
+              .toString('hex')}, error encountered when trying to add tx:\n${error}`
+          )
+        }
+      }
+      index++
+    }
+
+    // Stop builder if block isn't returned after 1 minute
+    setTimeout(() => {
+      this.stop(payloadId)
+    }, 60000)
+
+    return payloadId
+  }
+
+  /**
+   * Stops a pending payload
+   */
+  stop(payloadId: Buffer) {
+    const payload = this.pendingPayloads.find((p) => p[0].equals(payloadId))
+    if (!payload) return
+    void payload[1].revert()
+  }
+
+  /**
+   * Returns the completed block
+   */
+  async build(payloadId: Buffer): Promise<Block | void> {
```

Comment:
What is the difference in usage between this and the `start` method?  They both appear to be async functions that construct a block with the main difference being that `build` returns the block and `start` returns the payload id.

Reply:
this is from the engine API. `start` is called when the block should start being prepared, and returns an identifier to CL. Then when CL is ready for the block it will call `build` with the identifier.

Reply:
Got it.  And the section of code below in the `build` method just adds additional transactions to the already constructed block indicated by `payloadId` if it's not already full?

Reply:
that's just our own hacky implementation here, see the opening comment in the file:

```js
/**
 * In the future this class will build a pending block by keeping the
 * transaction set up-to-date with the state of local mempool until called.
 *
 * For now this simple implementation just adds txs from the pool when
 * started and called.
 */
```

Reply:
This implementation is really suboptimal because we should be shuffling txs to maximize MEV, not just adding new txs to the end of a non-full block, but since we are not targeting to be run as a tip of chain client it's less important to prioritize.